### DISCUSSION TITLE
Is it possible to have a hook each time PB connects to SQLite?

### DISCUSSION THREAD

**Discussion by flarco** [pocketbase/pocketbase#3656] [Q&A] [answered]
Hello, 

The reason I ask is that I have separate SQLite db files (some very large) that I `ATTACH` to the main PB SQLite, so I can join to those external tables directly. At the moment, each time I create a new query with `app.Dao().DB().NewQuery(sql)`, I run the command `attach database './external.db' as 'external'` right before, to ensure the external db is attached, since concurrent/separate connections are possible. Otherwise, I often get errors that it cannot find the `external` schema. Now this is not great design and slows down my avg API response time a bunch since it has to run it every time.

I looked but couldn't find, is there some hook like `onDbConnect()` that I could use to run this `attach` command only once each new connection?

**Comment by ganigeorgiev** [comment, marked as answer]
No, unfortunately there is no such hook at the moment and it will differ depending on whether CGO is enabled and which driver is being used (_see for example [db_cgo.go](https://github.com/pocketbase/pocketbase/blob/ea5ca009de839673d569da3d5950f8fcd4c59253/core/db_cgo.go#L24C4-L24C15); I'm not sure if such connect hook is available in the cgo-free port [modernc.org/sqlite](https://gitlab.com/cznic/sqlite)_).

There are plans to expose the above `core.connectDB` method to allow users registering their own SQLite configuration (eg. when using extensions) but it is left aside for now to implement together with the internals refactoring (hopefully final) after Presentator.

I don't know if it will perform better, but you can try to wrap your `ATTACH` calls in a condition that will send a dummy request to some of its table (eg. `select 1 from yourdb.someTable limit 1`). In case it fails -> perform the db attach, otherwise skip. The database should remains attached for the duration of the connection (idle connections are reused).

**Reply by ganigeorgiev** [reply]
_Actually you don't need the dummy select query, you should be able to get a list of the attached databases using [`PRAGMA database_list`](https://www.sqlite.org/pragma.html#pragma_database_list) and check for its existence. Also I'd expect the attach to throw an error when it is called multiple times with the same name for the connection (unless it was previously detached), aka. you can try to directly call your `attach` statement before the queries and just ignore/skip the "existing db" error (this is assuming that there is no impact on the performance when errorneous `ATTACH DATABASE` are executed)._

**Reply by flarco** [reply]
Yes, that is what I'm doing, right before the query, I'm running `_ = AttachCoreDb()` where any error is silently ignored. Still having slow API response, used to average about `30ms`, now is `120ms`... Likely it's something else, will analyze deeper. Thanks so much for your thoughts, much appreciated.
